Boost max flow algorithms do not compile

Boost provides three different algorithms for finding max flow in directed graphs: **boykov_kolmogorov**, **edmonds_karp** and **push_relabel**. All of them have named and non-named parameter versions. Parameter sets they use are also very similar.
Despite that, with same parameters some of these algorithms compile and some of them do not. This is the case only for graph without **edge_capacity_t** interior property.

**push_relabel** compiles nicely with both named and non-named version:
``` c++
 using Graph =
    boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS,
                        VertexProperty, EdgeProperty>;
  auto props = boost::capacity_map(capacity)
               .residual_capacity_map(residual_capacity)
               .reverse_edge_map(reverse_edge_map)
               .vertex_index_map(vertex_index_map)
               .color_map(color_map)
               .predecessor_map(predcessor_map)
               .distance_map(distance_map);
  boost::push_relabel_max_flow(g, s, t, props);
  boost::push_relabel_max_flow(g, s, t, capacity, residual_capacity,
                   reverse_edge_map, vertex_index_map);
```
**boykov_kolmogorov** compiles with non-named version:
```c++
  boost::boykov_kolmogorov_max_flow(g, capacity, residual_capacity,
                                    reverse_edge_map,
                                    vertex_index_map, s, t);
```
But fails with named version:
```c++
 boost::boykov_kolmogorov_max_flow(g, s, t, props);
```
> /celibs/boost_1_73_0/boost/graph/detail/adjacency_list.hpp:2768:17: error: forming reference to void

**edmonds_karp** fails with both named and non-named versions with same error:
```c++
boost::edmonds_karp_max_flow(g, s, t, props);
boost::edmonds_karp_max_flow(g, s, t, capacity, residual_capacity, reverse_edge_map,
                          color_map, predcessor_map);
```

> /celibs/boost_1_73_0/boost/concept_check.hpp:147:9: error: use of deleted function

This issue was also [asked on StackOverflow](https://stackoverflow.com/questions/64734193/boost-max-flow-algorithms-do-not-compile-error-forming-reference-to-void). User sehe suggested that the problem is connected to some subtle error in `choose_const_pmap` function

Full example:
[Live on godbolt](https://godbolt.org/z/dvjfec)
```c++
#include <boost/config.hpp>
#include <boost/graph/adjacency_list.hpp>
#include <boost/graph/graph_utility.hpp>
#include <boost/property_map/function_property_map.hpp>
#include <boost/graph/boykov_kolmogorov_max_flow.hpp>
#include <boost/graph/edmonds_karp_max_flow.hpp>
#include <boost/graph/push_relabel_max_flow.hpp>

int main(){ 
  struct VertexProperty final {};
  struct EdgeProperty final {};
  using Graph =
      boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS,
                            VertexProperty, EdgeProperty>;
  using Edge = boost::graph_traits<Graph>::edge_descriptor;
  using Vertex = boost::graph_traits<Graph>::vertex_descriptor;
  auto g = Graph{};
  auto s = Vertex{};
  auto t = Vertex{};

  auto residualCapacityMap = std::vector<int>{};
  auto reverseEdgeMap = std::vector<Edge>{};
  auto colorMap = std::vector<boost::default_color_type>{};
  auto predcessorMap = std::vector<Edge>{};
  auto distanceMap = std::vector<int>{};


  auto vertex_index_map = boost::make_function_property_map<Vertex>(
      [&](auto const &v) { return 0; });
  auto edge_index_map = boost::make_function_property_map<Edge>(
      [&](auto const &e) { return 0; });
  auto capacity = boost::make_function_property_map<Edge>(
      [&](auto const &e) -> int { return 0; });
  auto residual_capacity = boost::make_iterator_property_map(
      residualCapacityMap.begin(), edge_index_map);
  auto reverse_edge_map =
      boost::make_iterator_property_map(reverseEdgeMap.begin(), edge_index_map);
  auto color_map =
      boost::make_iterator_property_map(colorMap.begin(), vertex_index_map);
  auto predcessor_map = boost::make_iterator_property_map(predcessorMap.begin(),
                                                    vertex_index_map);
  auto distance_map =
      boost::make_iterator_property_map(distanceMap.begin(), vertex_index_map);

  auto props = boost::capacity_map(capacity)
                   .residual_capacity_map(residual_capacity)
                   .reverse_edge_map(reverse_edge_map)
                   .vertex_index_map(vertex_index_map)
                   .color_map(color_map)
                   .predecessor_map(predcessor_map)
                   .distance_map(distance_map);

  // push_relabel compiles nicely
  boost::push_relabel_max_flow(g, s, t, props);
  boost::push_relabel_max_flow(g, s, t, capacity, residual_capacity,
                       reverse_edge_map, vertex_index_map);

  // As well as boykov_kolmogorov_max_flow with non-named version
  boost::boykov_kolmogorov_max_flow(g, capacity, residual_capacity,
                                    reverse_edge_map,
                                    vertex_index_map, s, t);
  // But named version boykov_kolmogorov fails
#if 1
  boost::boykov_kolmogorov_max_flow(g, s, t, props);
#endif

  // Non-named and named versions of edmonds_karp fail with same error
#if 0
  boost::edmonds_karp_max_flow(g, s, t, props);
#endif
#if 0
  boost::edmonds_karp_max_flow(g, s, t, capacity, residual_capacity, reverse_edge_map,
                              color_map, predcessor_map);
#endif
}
```

--- a/include/boost/graph/boykov_kolmogorov_max_flow.hpp
+++ b/include/boost/graph/boykov_kolmogorov_max_flow.hpp
@@ -1029,8 +1029,7 @@ boykov_kolmogorov_max_flow(Graph& g, CapacityEdgeMap cap,
  * named-parameter version, some given
  */
 template < class Graph, class P, class T, class R >
-typename property_traits<
-    typename property_map< Graph, edge_capacity_t >::const_type >::value_type
+typename detail::edge_capacity_value< Graph, P, T, R >::type
 boykov_kolmogorov_max_flow(Graph& g,
     typename graph_traits< Graph >::vertex_descriptor src,
     typename graph_traits< Graph >::vertex_descriptor sink,

--- a/test/boykov_kolmogorov_max_flow_test.cpp
+++ b/test/boykov_kolmogorov_max_flow_test.cpp
@@ -188,10 +188,24 @@ long test_bundled_properties(int n_verts, int n_edges, std::size_t seed)
     boost::tie(src, sink)
         = fill_random_max_flow_graph(g, get(&tEdge::edge_capacity, g),
             get(&tEdge::edge_reverse, g), n_verts, n_edges, seed);
-    return boykov_kolmogorov_max_flow(g, get(&tEdge::edge_capacity, g),
-        get(&tEdge::edge_residual_capacity, g), get(&tEdge::edge_reverse, g),
-        get(&tVertex::vertex_predecessor, g), get(&tVertex::vertex_color, g),
-        get(&tVertex::vertex_distance, g), get(vertex_index, g), src, sink);
+
+    long flow_unnamed_overload = boykov_kolmogorov_max_flow(g,
+        get(&tEdge::edge_capacity, g), get(&tEdge::edge_residual_capacity, g),
+        get(&tEdge::edge_reverse, g), get(&tVertex::vertex_predecessor, g),
+        get(&tVertex::vertex_color, g), get(&tVertex::vertex_distance, g),
+        get(vertex_index, g), src, sink);
+
+    long flow_named_overload = boykov_kolmogorov_max_flow(g, src, sink,
+        capacity_map(get(&tEdge::edge_capacity, g))
+            .residual_capacity_map(get(&tEdge::edge_residual_capacity, g))
+            .reverse_edge_map(get(&tEdge::edge_reverse, g))
+            .predecessor_map(get(&tVertex::vertex_predecessor, g))
+            .color_map(get(&tVertex::vertex_color, g))
+            .distance_map(get(&tVertex::vertex_distance, g))
+            .vertex_index_map(get(vertex_index, g)));
+
+    BOOST_TEST(flow_unnamed_overload == flow_named_overload);
+    return flow_named_overload;
 }
 
 long test_overloads(int n_verts, int n_edges, std::size_t seed)

--- a/test/boykov_kolmogorov_max_flow_test.cpp
+++ b/test/boykov_kolmogorov_max_flow_test.cpp
@@ -72,7 +72,7 @@ fill_random_max_flow_graph(Graph& g, CapacityMap cap, ReverseEdgeMap rev,
     // cannot use this, as we have no idea how properties are stored, right?
     typename graph_traits< Graph >::edge_iterator ei, e_end;
     for (boost::tie(ei, e_end) = edges(g); ei != e_end; ++ei)
-        cap[*ei] = int_gen();
+        put(cap, *ei, int_gen());
 
     // get source and sink node
     vertex_descriptor s = random_vertex(g, gen);
@@ -96,9 +96,9 @@ fill_random_max_flow_graph(Graph& g, CapacityMap cap, ReverseEdgeMap rev,
         boost::tie(new_edge, inserted)
             = add_edge(source_vertex, target_vertex, g);
         assert(inserted);
-        rev[old_edge] = new_edge;
-        rev[new_edge] = old_edge;
-        cap[new_edge] = 0;
+        put(rev, old_edge, new_edge);
+        put(rev, new_edge, old_edge);
+        put(cap, new_edge, 0);
     }
     return std::make_pair(s, t);
 }

